In [1]:
#importing the tools I need
import numpy as np
import pandas as pd

from pandas import DataFrame

In [2]:
#uploaded my data from my GitHub into this Juypter notebook
aflights = pd.read_csv("https://raw.githubusercontent.com/rleon207/DAV5400/master/CSV_A10.csv")

In [3]:
#sanity check
aflights

Unnamed: 0 Unnamed: 1  Los Angeles  Phoenix  San Diego  San Fran  Seattle
0     ALASKA    on time        497.0    221.0      212.0     503.0   1841.0
1        NaN    delayed         62.0     12.0       20.0     102.0    305.0
2        NaN        NaN          NaN      NaN        NaN       NaN      NaN
3    AM WEST    on time        694.0   4840.0      383.0     320.0    201.0
4        NaN    delayed        117.0    415.0       65.0     129.0     61.0

In [4]:
#changed column names
aflights = aflights.rename(columns={"Unnamed: 0": "Airline", "Unnamed: 1": "Status"})
aflights

Airline   Status  Los Angeles  Phoenix  San Diego  San Fran  Seattle
0   ALASKA  on time        497.0    221.0      212.0     503.0   1841.0
1      NaN  delayed         62.0     12.0       20.0     102.0    305.0
2      NaN      NaN          NaN      NaN        NaN       NaN      NaN
3  AM WEST  on time        694.0   4840.0      383.0     320.0    201.0
4      NaN  delayed        117.0    415.0       65.0     129.0     61.0

Think about how you would want the data to appear if it were converted to "long" format?
- I think I would want the data to only have four columns, and the last column would have all of the unique frequency values for each observation.

How would you define a "single observation" for the data shown in the graphic?
- A single observation would be the unique minute value for any given airline, status, and city. For example, Alaska's minute count for delayed flights in Phoenix - one observation.

How many key values are associated with each data value?
- Three key values are associated with each data value, airline, status, and city.

How many columns should your long format structure contain based on the information provided in the graphic shown above?
- My long format structure should have no more than four columns based on the graphic shown above.

What would the column headings for the long structure be?
- My column headings should be as follows: Airline, Status, City, and Frequency.

In [5]:
#dropped the NaN row
aflights = aflights.drop(aflights.index[2])
aflights

Airline   Status  Los Angeles  Phoenix  San Diego  San Fran  Seattle
0   ALASKA  on time        497.0    221.0      212.0     503.0   1841.0
1      NaN  delayed         62.0     12.0       20.0     102.0    305.0
3  AM WEST  on time        694.0   4840.0      383.0     320.0    201.0
4      NaN  delayed        117.0    415.0       65.0     129.0     61.0

In [6]:
#replaced the "NaN" values with the appropriate Airline
aflights['Airline'].fillna(method='ffill', inplace = True)
aflights

Airline   Status  Los Angeles  Phoenix  San Diego  San Fran  Seattle
0   ALASKA  on time        497.0    221.0      212.0     503.0   1841.0
1   ALASKA  delayed         62.0     12.0       20.0     102.0    305.0
3  AM WEST  on time        694.0   4840.0      383.0     320.0    201.0
4  AM WEST  delayed        117.0    415.0       65.0     129.0     61.0

In [7]:
#used melt function to make the wanted columns into variables, all of my minutes turned into one column
longflights = aflights.melt(id_vars = ['Airline', 'Status'],  value_vars = ['Los Angeles', 'Phoenix', 'San Diego', 'San Fran', 'Seattle'])
longflights


Airline   Status     variable   value
0    ALASKA  on time  Los Angeles   497.0
1    ALASKA  delayed  Los Angeles    62.0
2   AM WEST  on time  Los Angeles   694.0
3   AM WEST  delayed  Los Angeles   117.0
4    ALASKA  on time      Phoenix   221.0
5    ALASKA  delayed      Phoenix    12.0
6   AM WEST  on time      Phoenix  4840.0
7   AM WEST  delayed      Phoenix   415.0
8    ALASKA  on time    San Diego   212.0
9    ALASKA  delayed    San Diego    20.0
10  AM WEST  on time    San Diego   383.0
11  AM WEST  delayed    San Diego    65.0
12   ALASKA  on time     San Fran   503.0
13   ALASKA  delayed     San Fran   102.0
14  AM WEST  on time     San Fran   320.0
15  AM WEST  delayed     San Fran   129.0
16   ALASKA  on time      Seattle  1841.0
17   ALASKA  delayed      Seattle   305.0
18  AM WEST  on time      Seattle   201.0
19  AM WEST  delayed      Seattle    61.0

In [8]:
#minor adjustment to have columns with relevant names
longflights = longflights.rename(columns={"variable": "City", "value": "Frequency"})
longflights

Airline   Status         City  Frequency
0    ALASKA  on time  Los Angeles      497.0
1    ALASKA  delayed  Los Angeles       62.0
2   AM WEST  on time  Los Angeles      694.0
3   AM WEST  delayed  Los Angeles      117.0
4    ALASKA  on time      Phoenix      221.0
5    ALASKA  delayed      Phoenix       12.0
6   AM WEST  on time      Phoenix     4840.0
7   AM WEST  delayed      Phoenix      415.0
8    ALASKA  on time    San Diego      212.0
9    ALASKA  delayed    San Diego       20.0
10  AM WEST  on time    San Diego      383.0
11  AM WEST  delayed    San Diego       65.0
12   ALASKA  on time     San Fran      503.0
13   ALASKA  delayed     San Fran      102.0
14  AM WEST  on time     San Fran      320.0
15  AM WEST  delayed     San Fran      129.0
16   ALASKA  on time      Seattle     1841.0
17   ALASKA  delayed      Seattle      305.0
18  AM WEST  on time      Seattle      201.0
19  AM WEST  delayed      Seattle       61.0

Data is now tidy.

In [9]:
#making an Alaska only DF
Alaska = (longflights.loc[longflights['Airline'] == 'ALASKA'])
Alaska

Airline   Status         City  Frequency
0   ALASKA  on time  Los Angeles      497.0
1   ALASKA  delayed  Los Angeles       62.0
4   ALASKA  on time      Phoenix      221.0
5   ALASKA  delayed      Phoenix       12.0
8   ALASKA  on time    San Diego      212.0
9   ALASKA  delayed    San Diego       20.0
12  ALASKA  on time     San Fran      503.0
13  ALASKA  delayed     San Fran      102.0
16  ALASKA  on time      Seattle     1841.0
17  ALASKA  delayed      Seattle      305.0

In [10]:
#number of times Alaska airport was on time regardless of city
(Alaska[Alaska['Status'] == 'on time']['Frequency'].sum())

3274.0

Alaska had 3,274 on time flights, regardless of city.

In [11]:
#ratio of Alaska's on time flights versus all flights
(Alaska[Alaska['Status'] == 'on time']['Frequency'].sum())/(Alaska['Frequency'].sum())

0.8672847682119206

Out of all of Alaska's flights in our dataset, Alaska is on time roughly 87% of the time.

In [12]:
#making an AmWest only DF
AmWest = (longflights.loc[longflights['Airline'] == 'AM WEST'])
AmWest

Airline   Status         City  Frequency
2   AM WEST  on time  Los Angeles      694.0
3   AM WEST  delayed  Los Angeles      117.0
6   AM WEST  on time      Phoenix     4840.0
7   AM WEST  delayed      Phoenix      415.0
10  AM WEST  on time    San Diego      383.0
11  AM WEST  delayed    San Diego       65.0
14  AM WEST  on time     San Fran      320.0
15  AM WEST  delayed     San Fran      129.0
18  AM WEST  on time      Seattle      201.0
19  AM WEST  delayed      Seattle       61.0

In [13]:
#AmWest's total flights that were on time.
(AmWest[AmWest['Status'] == 'on time']['Frequency'].sum())

6438.0

AmWest had 6,438 on time flights, regardless of city.

In [14]:
#AmWest's ratio of on time flights over total flights.
(AmWest[AmWest['Status'] == 'on time']['Frequency'].sum())/(AmWest['Frequency'].sum())

0.8910726643598615

Out of all of Alaska's flights in our dataset, AmWest is on time roughly 89% of the time.
This proves to be a better ratio, even given AmWest's greater quantity of flight demand. AmWest is a more punctual airline in both capacities.

If I were to change my data back to a "wide" format, I would try to mimic the structure of the orignal graph I started with. It is a very simple, easy-to-read, intuitive table that wouldn't need much explanation.

In [15]:
#showing the original long DF for viewing.
longflights

Airline   Status         City  Frequency
0    ALASKA  on time  Los Angeles      497.0
1    ALASKA  delayed  Los Angeles       62.0
2   AM WEST  on time  Los Angeles      694.0
3   AM WEST  delayed  Los Angeles      117.0
4    ALASKA  on time      Phoenix      221.0
5    ALASKA  delayed      Phoenix       12.0
6   AM WEST  on time      Phoenix     4840.0
7   AM WEST  delayed      Phoenix      415.0
8    ALASKA  on time    San Diego      212.0
9    ALASKA  delayed    San Diego       20.0
10  AM WEST  on time    San Diego      383.0
11  AM WEST  delayed    San Diego       65.0
12   ALASKA  on time     San Fran      503.0
13   ALASKA  delayed     San Fran      102.0
14  AM WEST  on time     San Fran      320.0
15  AM WEST  delayed     San Fran      129.0
16   ALASKA  on time      Seattle     1841.0
17   ALASKA  delayed      Seattle      305.0
18  AM WEST  on time      Seattle      201.0
19  AM WEST  delayed      Seattle       61.0

In [19]:
#making an array to help reshape my df back to wide.
array = ['Airline', 'Status']

In [21]:
#my data now matches the original wide form.
longflights.pivot_table(columns ='City', index=array)

Frequency                                   
City            Los Angeles Phoenix San Diego San Fran Seattle
Airline Status                                                
ALASKA  delayed        62.0    12.0      20.0    102.0   305.0
        on time       497.0   221.0     212.0    503.0  1841.0
AM WEST delayed       117.0   415.0      65.0    129.0    61.0
        on time       694.0  4840.0     383.0    320.0   201.0